File: v1/src/main/java/com/api/v1/utils/BookFinderUtil.java
package com.api.v1.utils;

import com.api.v1.domain.entities.Book;
import com.api.v1.domain.repositories.BookRepository;
import com.api.v1.exceptions.book.BookNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class BookFinderUtil {

    @Autowired
    private BookRepository repository;

    public Mono<Book> find(String isbn) {
        return repository
                .findAll()
                .filter(e -> e.getIsbn().equals(isbn)
                        && e.getArchivedAt() == null
                )
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new BookNotFoundException(isbn)));
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/book/utils/BookFinderUtil.kt
package com.api.v2.book.utils

import com.api.v2.book.exceptions.BookNotFoundException
import com.api.v2.book.domain.Book
import com.api.v2.book.domain.BookRepository
import com.api.v2.exceptions.EmptyEntityException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.count
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.singleOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class BookFinderUtil {

    @Autowired
    private lateinit var bookRepository: BookRepository

    suspend fun findOne(isbn: String): Book {
        return withContext(Dispatchers.IO) {
            val existingBook = bookRepository
                    .findAll()
                    .filter { e -> e.isbn == isbn && e.archivedAt == null }
                    .singleOrNull()
            if (existingBook == null) throw BookNotFoundException(isbn)
            existingBook
        }
    }

    suspend fun findMany(isbn: String): Flow<Book> {
        return withContext(Dispatchers.IO) {
            val books = bookRepository
                .findAll()
                .filter { e -> e.isbn == isbn }
            if (books.count() == 0) throw EmptyEntityException()
            books
        }
    }

}